File: 20190519python/0257. Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        if not root:
            return []
        rep=[]
        self.getBinaryTreePaths(root,str(root.val),rep)
        return rep
    
    def getBinaryTreePaths(self,root,cur,rep):
        if not root.left and not root.right:
            rep.append(cur)
        elif root.left and not root.right:
            self.getBinaryTreePaths(root.left,cur+"->"+str(root.left.val),rep)
        elif not root.left and root.right:
            self.getBinaryTreePaths(root.right,cur+"->"+str(root.right.val),rep)
        else:
            self.getBinaryTreePaths(root.left,cur+"->"+str(root.left.val),rep)
            self.getBinaryTreePaths(root.right,cur+"->"+str(root.right.val),rep)
        

File: archives/C#/0257. Binary Tree Paths.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<string> BinaryTreePaths(TreeNode root) {
        IList<string> rep=new List<string>();
        if(root==null){
            return rep;
        }
        Queue<string> strQueue=new Queue<string>();
        Queue<TreeNode> rootQueue=new Queue<TreeNode>();
        strQueue.Enqueue(root.val.ToString());
        rootQueue.Enqueue(root);
        while(rootQueue.Count()!=0){
            int lenRootQueue=rootQueue.Count();
            for (int i=0;i<lenRootQueue;i++){
                string curString=strQueue.Dequeue();
                TreeNode curRoot=rootQueue.Dequeue();
                if(curRoot.left==null && curRoot.right==null){
                    rep.Add(curString);
                }
                if(curRoot.left!=null){
                    rootQueue.Enqueue(curRoot.left);
                    strQueue.Enqueue(curString+"->"+curRoot.left.val.ToString());
                }
                if(curRoot.right!=null){
                    rootQueue.Enqueue(curRoot.right);
                    strQueue.Enqueue(curString+"->"+curRoot.right.val.ToString());
                }
            }
        }
        return rep;
    }
}
